File: core/src/main/scala/io/github/mathter/zi/dsl/ListSource.scala
package io.github.mathter.zi.dsl

import scala.reflect.ClassTag

trait ListSource[T] extends Source[List[T]] {
  def etp: ClassTag[T]

  def first: Source[T]

  def last: Source[T]

  def index(index: Source[Int]): Source[T]

  def mapElem[D](f: T => D)(implicit ctag: ClassTag[D]): ListSource[D]
}

File: jcore/src/main/java/io/github/mathter/jzi/dsl/ListSource.java
package io.github.mathter.jzi.dsl;

import io.github.mathter.zi.dsl.Source;

import java.util.List;
import java.util.function.Function;

public interface ListSource<T> extends Source<List<T>> {
    public Source<T> first();

    public Source<T> last();

    public Source<T> index(Source<Integer> index);

    public <D> ListSource<D> mapElem(Function<? super T, ? extends D> mapper);

    public <D> ListSource<D> mapsElem(Function<Source<T>, Source<D>> mapper);

    public <K> Group<K, T> group(Function<Source<T>, Source<K>> keyMapper);
}
